entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -289,20 +290,126 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
       AttributeValue attributeValue =
           EntityQueryConverter.convertToAttributeValue(setAttribute.getValue()).build();
       String jsonValue = DocStoreJsonFormat.printer().print(attributeValue);
+      JSONDocument jsonDocument = new JSONDocument(jsonValue);
 
+      Map<Key, Map<String, Document>> entitiesUpdateMap = new HashMap<>();
       for (String entityId : request.getEntityIdsList()) {
         SingleValueKey key =
             new SingleValueKey(requestContext.getTenantId().orElseThrow(), entityId);
         // TODO better error reporting once doc store exposes the,
-        if (!entitiesCollection.updateSubDoc(key, subDocPath, new JSONDocument(jsonValue))) {
-          LOG.warn(
-              "Failed to update entity {}, subDocPath {}, with new doc {}.",
-              key,
-              subDocPath,
-              jsonValue);
+        if (entitiesUpdateMap.containsKey(key)) {
+          entitiesUpdateMap.get(key).put(subDocPath, new JSONDocument(jsonValue));
+        } else {
+          Map<String, Document> subDocument = new HashMap<>();
+          subDocument.put(subDocPath, new JSONDocument(jsonValue));
+          entitiesUpdateMap.put(key, subDocument);
         }
       }
+      try {
+        entitiesCollection.bulkUpdateSubDocs(entitiesUpdateMap);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to update entities, subDocPath {}, with new doc {}.", subDocPath, jsonValue, e);
+        throw e;
+      }
+    }
+  }
+
+  @Override
+  public void bulkUpdate(
+      BulkEntityUpdateRequest request, StreamObserver<ResultSetChunk> responseObserver) {
+    // Validations
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Optional<String> maybeTenantId = requestContext.getTenantId();
+    if (maybeTenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+    if (StringUtils.isEmpty(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Entity type is missing in the request."));
+      return;
+    }
+    if (request.getEntitiesCount() == 0) {
+      responseObserver.onError(new ServiceException("Entities are missing in the request."));
+    }
+    String tenantId = maybeTenantId.get();
```

Comment:
Please pass in the `tenantId` directly to `doBulkUpdate`, instead of the `requestContext`

Reply:
nit: an extra line above

Reply:
doBulkUpdate calls the method transformUpdateOperations () which needs requestContext to find the subDocPath.
`String subDocPath =
          entityAttributeMapping
              .getDocStorePathByAttributeId(requestContext, attributeId)
              .orElseThrow(
                  () -> new IllegalArgumentException("Unknown attribute FQN " + attributeId));`

Reply:
Yeah, agree with keeping it request context - we really want to keep it as a context object until the last second, tenant IDs should only be resolved within the method that uses them.